### PR TITLE
feat: add an ability to delete a group gf-80

### PR DESCRIPTION
**Pull request by kolibri753**
1. Options Icon Addition:
image
Added an options icon as a new column at the end of each row in the Groups table.

2.  Options Group Menu:
image
Implemented an options group menu that appears on clicking the icon. Added a new popover hook, usePosition, to calculate where the menu should appear. This hook is optional and ensures that the menu’s appearance remains consistent, even in projects with absolute positioning.

3. Confirmation Modal:
image
Added a confirmation modal that appears when the delete option is selected. The modal's styling was adjusted to match the Figma design specifications.

4. Group Deletion:
image
image
image

Implemented backend functionality to delete a group upon confirmation. Updated the table and adjusted pagination to reflect changes when a group is deleted.

### CODE REVIEW

File: apps/backend/src/modules/groups/group.controller.ts
Comment (yevhenii-hladkov) [inline apps/backend/src/modules/groups/group.controller.ts:150]: Since `id` is already typed as a number, there's no need to convert it to a Number.

File: apps/backend/src/modules/groups/group.controller.ts
Comment (yevhenii-hladkov) [inline apps/backend/src/modules/groups/group.controller.ts:153]:
```suggestion
			payload: await this.groupService.delete(id),
```

Comment (yevhenii-hladkov) [reply apps/backend/src/modules/groups/group.controller.ts:153]: Actually, you can leave it as is. I missed something

Comment (kolibri753) [reply apps/backend/src/modules/groups/group.controller.ts:153]: It's just a draft, I haven't thought much into it yet, thanks for advices :cat2:

File: apps/backend/src/modules/groups/group.controller.ts
Comment (GvoFor) [inline apps/backend/src/modules/groups/group.controller.ts:69]: The method should be "DELETE"

Comment (kolibri753) [reply apps/backend/src/modules/groups/group.controller.ts:69]: thanks, I've already fixed it, just haven't pushed yet (electricity)

File: apps/backend/src/modules/groups/group.service.ts
Comment (GvoFor) [inline apps/backend/src/modules/groups/group.service.ts:57]: Do we really need this check? Maybe lets just return `true` if group was deleted and `false` otherwise?

Comment (GvoFor) [reply apps/backend/src/modules/groups/group.service.ts:57]: At least, you can avoid `find` request, and call `delete` right away and then check if result is `false` (item was not deleted), that means that there is no item with this id and we can throw error

File: apps/frontend/src/pages/access-management/libs/components/group-popover/group-popover.tsx
Comment (kolibri753) [inline apps/frontend/src/pages/access-management/libs/components/group-popover/group-popover.tsx:47]: Menu component?

File: apps/frontend/src/assets/images/icons/pen.svg
Comment (GvoFor) [inline apps/frontend/src/assets/images/icons/pen.svg:1]: Let's call this icon `pencil`

File: apps/frontend/src/assets/images/icons/trash.svg
Comment (GvoFor) [inline apps/frontend/src/assets/images/icons/trash.svg:1]: Let's call this icon `trash-can`

File: apps/frontend/src/assets/images/icons/pen.svg
Comment (kolibri753) [reply apps/frontend/src/assets/images/icons/pen.svg:1]: not sure about this, I see that some of us used "edit"

Comment (kolibri753) [reply apps/frontend/src/assets/images/icons/pen.svg:1]: I was thinking of waiting until the first one is merged, and then redoing it to match. Or would it be better if we discussed this first?

File: apps/backend/src/modules/groups/group.service.ts
Comment (kolibri753) [inline apps/backend/src/modules/groups/group.service.ts:59]: Maybe I should just return await this.groupRepository.delete(id); instead of this.

File: apps/frontend/src/modules/groups/slices/actions.ts
Comment (AnnKasian) [inline apps/frontend/src/modules/groups/slices/actions.ts:19]: Maybe just `delete`? We  don't have another version of deleting, just by Id

File: apps/backend/src/modules/groups/group.repository.ts
Comment (kolibri753) [reply apps/backend/src/modules/groups/group.repository.ts:41]: Yeah, I thought about it and even checked other PRs later, and found that they [exported](https://github.com/BinaryStudioAcademy/bsa-2024-gitfit/pull/168#discussion_r1742173332) it. So I'll change it.

File: apps/frontend/src/modules/groups/slices/actions.ts
Comment (kolibri753) [reply apps/frontend/src/modules/groups/slices/actions.ts:19]: If shortly, it's a reserved keyword

File: apps/frontend/src/assets/images/icons/pen.svg
Comment (GvoFor) [reply apps/frontend/src/assets/images/icons/pen.svg:1]: >I see that some of us used "edit"

I asked her to rename it to `pencil` too (she already done this)
I also asked her to rename `options` to `ellipsis`

The point (in my opinion) is that icon name should represent its content precisely. For example, from `edit` name can you imagine, how the icon looks like? or from `options` name?

File: apps/backend/src/modules/groups/group.service.ts
Comment (GvoFor) [reply apps/backend/src/modules/groups/group.service.ts:59]: I clarified this and you should keep this as it is. Check [this](https://bsa2024.slack.com/archives/C07HWM8616U/p1725378501043329).

File: apps/frontend/src/assets/images/icons/pen.svg
Comment (kolibri753) [reply apps/frontend/src/assets/images/icons/pen.svg:1]: It can be the opposite too. Sometimes we know what the icon represents (like edit or delete), but the design isn’t final. Using a generic name allows us to swap the icon without renaming everything.

Comment (GvoFor) [reply apps/frontend/src/assets/images/icons/pen.svg:1]: Hm. That's a good point. Then lets @liza-veis decide

Comment (GvoFor) [reply apps/frontend/src/assets/images/icons/pen.svg:1]: By the way, our design is final (at least this right for icons)

Comment (kolibri753) [reply apps/frontend/src/assets/images/icons/pen.svg:1]: I just wanted to tell the opposite opinion 😅

**Comment by kolibri753** [comment]
Working on fixing that

1. with position relative on options container
The popup is displayed correctly, even with a scrollbar
image

however, the last popup gets rendered inside the table
image

2. without position relative on options container
The popup is displayed correctly when there is no scrollbar

But, when there are many elements and the table overflows, the popup does not adjust properly.
image

**Comment by Fjortis** [comment]
> Working on fixing that
> 
> 1. with position relative on options container
>    The popup is displayed correctly, even with a scrollbar
>    image
> 
> however, the last popup gets rendered inside the table image
> 
> 2. without position relative on options container
>    The popup is displayed correctly when there is no scrollbar
> 
> But, when there are many elements and the table overflows, the popup does not adjust properly. image
what i think of is to try to use position: absolute with a higher level container, or use position absolutefor popup and dynmically adjust positioning.

File: apps/frontend/src/libs/components/confirmation-modal/confirmation-modal.tsx
Comment (perekotypole) [reply apps/frontend/src/libs/components/confirmation-modal/confirmation-modal.tsx:25]: We can't use simple functions for events

File: apps/backend/src/modules/groups/group.controller.ts
Comment (kolibri753) [reply apps/backend/src/modules/groups/group.controller.ts:161]: *status: HTTPCode.OK,

**Comment by kolibri753** [comment]
Should we freeze the table from scrolling when the popover menu is opened?

File: apps/frontend/src/libs/components/menu/menu.tsx
Comment (what1s1ove) [inline apps/frontend/src/libs/components/menu/menu.tsx:10]:
```suggestion
	usePositioning?: boolean;
```
`use` is a better prefix for functions. use `has` instead

File: apps/frontend/src/libs/components/popover/popover.tsx
Comment (what1s1ove) [inline apps/frontend/src/libs/components/popover/popover.tsx:14]: same

**Review by what1s1ove — APPROVED** [review]

File: apps/frontend/src/libs/components/menu/menu.tsx
Comment (kolibri753) [reply apps/frontend/src/libs/components/menu/menu.tsx:10]: Maybe `enablePositioning` is better than `hasPositioning`? Has sounds like it's already active, while enable makes it clear we’re deciding to turn it on or not.

File: apps/frontend/src/pages/access-management/libs/components/groups-table/groups-table.tsx
Comment (liza-veis) [inline apps/frontend/src/pages/access-management/libs/components/groups-table/groups-table.tsx:44]: It looks like we only need id, not the whole group

File: apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts
Comment (liza-veis) [inline apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts:11]: do we need this flag at all? can't we use the same logic for all popups?
```suggestion
```

File: apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts
Comment (liza-veis) [inline apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts:10]: What is isActive? if it means if popup is open, then let's name it isOpen

File: apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts
Comment (liza-veis) [inline apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts:23]: Can't we use just one of these values?

File: apps/frontend/src/modules/groups/slices/group.slice.ts
Comment (liza-veis) [inline apps/frontend/src/modules/groups/slices/group.slice.ts:40]:
```suggestion
			state.groupsTotalCount -= GROUPS_DELETE_COUNT;
```

File: apps/frontend/src/libs/components/table/table.tsx
Comment (liza-veis) [inline apps/frontend/src/libs/components/table/table.tsx:61]: no, we need to do this differently. The table component should now nothing about options and what if we will have more than just one options column? One of the possible solutions to pass width as an optional parameter to column config
```suggestion
```

File: apps/frontend/src/libs/components/table/table.tsx
Comment (liza-veis) [inline apps/frontend/src/libs/components/table/table.tsx:69]: What is this for?

**Review by liza-veis — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/components/menu/menu.tsx
Comment (what1s1ove) [reply apps/frontend/src/libs/components/menu/menu.tsx:10]: prefer `has`/`is`

File: apps/frontend/src/libs/components/table/table.tsx
Comment (perekotypole) [reply apps/frontend/src/libs/components/table/table.tsx:69]: `flexRender` doesn't display react components, so we have to check what the content of cell

File: apps/frontend/src/libs/components/icon-button/styles.module.css
Comment (kolibri753) [reply apps/frontend/src/libs/components/icon-button/styles.module.css:2]: without it we are not able to open popover menu on project

File: apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts
Comment (kolibri753) [reply apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts:11]: we use absolute/relative positioning in projects module. If we don't use it, we are allowed to use this hook, but the full width of the project won't be clickable

File: apps/frontend/src/pages/access-management/libs/components/groups-table/groups-table.tsx
Comment (kolibri753) [reply apps/frontend/src/pages/access-management/libs/components/groups-table/groups-table.tsx:53]: not sure I understand correctly, smth like this?
```
const groupColumns = getGroupColumns({
		onDelete: (groupId: number) => {
			handleDeleteClick({
				groupId,
				groups,
				onOpen: onDeleteModalOpen,
				setGroupIdToDelete,
			});
		},
	});
```

Comment (kolibri753) [reply apps/frontend/src/pages/access-management/libs/components/groups-table/groups-table.tsx:53]: or just
```
	const onDelete = useCallback(
		(groupId: number) => {
			const groupExists = groups.some(({ id }) => id === groupId);

			if (groupExists) {
				setGroupIdToDelete(groupId);
				onDeleteModalOpen();
			}
		},
		[groups, onDeleteModalOpen],
	);

	const groupColumns = getGroupColumns(onDelete);
```

Comment (kolibri753) [reply apps/frontend/src/pages/access-management/libs/components/groups-table/groups-table.tsx:53]: I think it's better to delete unnecessary checks and just leave this
```
const onDelete = useCallback(
	(groupId: number) => {
		setGroupId(groupId);
		onDeleteModalOpen();
	},
	[onDeleteModalOpen],
);

```

File: apps/backend/src/modules/groups/group.repository.ts
Comment (GvoFor) [inline apps/backend/src/modules/groups/group.repository.ts:54]:
```suggestion
		const group = await this.groupModel.query().findById(id).execute();
```

File: apps/frontend/src/modules/groups/slices/group.slice.ts
Comment (GvoFor) [reply apps/frontend/src/modules/groups/slices/group.slice.ts]: Do we really need to keep track `groupsTotalCount` in state? Can't we use `groups.length` instead? Or this may lead to unwanted rerenders?

File: apps/frontend/src/pages/access-management/libs/components/group-menu/group-menu.tsx
Comment (GvoFor) [inline apps/frontend/src/pages/access-management/libs/components/group-menu/group-menu.tsx:19]:
```suggestion
		<div className={styles["menu-container"]}>
```

File: apps/frontend/src/modules/groups/slices/group.slice.ts
Comment (perekotypole) [reply apps/frontend/src/modules/groups/slices/group.slice.ts]: @Fjortis It has an effect on table pagination. `groups.length` will only show the number of elements on a page.

File: apps/frontend/src/pages/access-management/libs/components/group-menu/group-menu.tsx
Comment (liza-veis) [inline apps/frontend/src/pages/access-management/libs/components/group-menu/group-menu.tsx:11]:
```suggestion
const GroupMenu = ({ groupId, onDelete }: Properties): JSX.Element => {
```

File: apps/frontend/src/pages/access-management/access-management.tsx
Comment (liza-veis) [inline apps/frontend/src/pages/access-management/access-management.tsx:174]:
```suggestion
			{hasSelectedGroup && (
```

File: apps/frontend/src/pages/access-management/access-management.tsx
Comment (liza-veis) [inline apps/frontend/src/pages/access-management/access-management.tsx:77]:
```suggestion
	const [selectedGroupId, setSelectedGroupId] = useState<null | number>(null);
```

File: apps/frontend/src/modules/groups/slices/group.slice.ts
Comment (liza-veis) [inline apps/frontend/src/modules/groups/slices/group.slice.ts:32]: As we need to handle pagination, it's too complicated to imitate it in reducer. Instead of this let's just do the same trick as when we close modals,  and call onPageSet with the initial page value on success, which will trigger the request and update the group. We want to keep the same pageSize and update page to the initial value.

```suggestion
```

File: apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts
Comment (liza-veis) [inline apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts:22]: I meant to use either `window.scrollY` or `document.documentElement.scrollTop` only in my first question

File: apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts
Comment (liza-veis) [inline apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts:19]: I think we don't need to change positioning depending on isOpen (we will just not display it).
```suggestion
		if (!hasPositioning || !reference.current) {
			return;
		}
```

File: apps/frontend/src/libs/components/popover/popover.tsx
Comment (liza-veis) [inline apps/frontend/src/libs/components/popover/popover.tsx:30]: let's rename it to hasAbsolutePositioning or hasFixedPositioning

**Review by liza-veis — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/access-management/libs/helpers/get-group-columns.helper.tsx
Comment (kolibri753) [reply apps/frontend/src/pages/access-management/libs/helpers/get-group-columns.helper.tsx:28]: a non-string header (like an empty string), the column requires an id

File: apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts
Comment (kolibri753) [reply apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts:21]: I prefer to leave both, but I can choose one (window.scrollY modern option)

File: apps/frontend/src/modules/groups/slices/group.slice.ts
Comment (kolibri753) [reply apps/frontend/src/modules/groups/slices/group.slice.ts]: I deleted the pagination code from the slice (except for `groupsTotalCount`), and it is working fine without any additional changes. Should I still modify this?

File: apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts
Comment (liza-veis) [reply apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts:21]: I think we need to leave document.documentElement.scrollTop only, it should cover all the cases

Comment (kolibri753) [reply apps/frontend/src/libs/hooks/use-popover-position/use-popover-position.hook.ts:21]: noted

File: apps/frontend/src/pages/access-management/access-management.tsx
Comment (liza-veis) [inline apps/frontend/src/pages/access-management/access-management.tsx:109]: we can remove groups/users dispatch inside this action
```suggestion
					id: selectedGroupId as number,
```
I want it to be done like this

```tsx
	useEffect(() => {
		if (groupDeleteStatus === DataStatus.FULFILLED) {
			onGroupPageChange(INITIAL_PAGE);
	        onUserPageChange(INITIAL_PAGE);
		}
	}, [projectPatchStatus, onUserPageChange, onGroupPageChange]);
```

File: apps/frontend/src/pages/access-management/access-management.tsx
Comment (perekotypole) [inline apps/frontend/src/pages/access-management/access-management.tsx:127]: Shouldn't it be `handleDelete`? (A2)

File: apps/frontend/src/pages/access-management/access-management.tsx
Comment (kolibri753) [reply apps/frontend/src/pages/access-management/access-management.tsx:109]: Sorry, I'm having trouble understanding what `INITIAL_PAGE` refers to. If it’s just 1, the implementation is simple, but after deletion, there's an issue with pagination. I can investigate it further, but wouldn't it be better to stay on the current page where the action was performed, and only move to the previous page if the deleted group was the last one on that page?

Comment (liza-veis) [reply apps/frontend/src/pages/access-management/access-management.tsx]: agree with you, let's leave it as you currently done it

File: apps/frontend/src/pages/access-management/libs/components/group-menu/group-menu.tsx
Comment (liza-veis) [reply apps/frontend/src/pages/access-management/libs/components/group-menu/group-menu.tsx:11]: this comment is still not fixed

**Review by liza-veis — APPROVED** [review]

**Pull request merged**
